### ISSUE TITLE
allowDiskUse Flag for large collections

### ISSUE THREAD

**Issue by clteich** [scullxbones/akka-persistence-mongo#317]
Hello,

after migrating about 5 millions documents from an old MongoDB Instance to our new MongoDB Atlas Cluster (M30 cluster tier), I get the following error

```
com.mongodb.MongoCommandException: Command failed with error 16945 (Location16945): 'Exceeded memory limit for $group, but didn't allow external sort. Pass allowDiskUse:true to opt in.' on server payback-prelive-shard-00-01-xfdte.mongodb.net:27017. The full response is {\"operationTime\": {\"$timestamp\": {\"t\": 1584450715, \"i\": 1}}, \"ok\": 0.0, \"errmsg\": \"Exceeded memory limit for $group, but didn't allow external sort. Pass allowDiskUse:true to opt in.\", \"code\": 16945, \"codeName\": \"Location16945\", \"$clusterTime\": {\"clusterTime\": {\"$timestamp\": {\"t\": 1584450715, \"i\": 1}}, \"signature\": {\"hash\": {\"$binary\": \"8tRyuPTZUcPgbsjt2BrwVM0ei58=\", \"$type\": \"00\"}, \"keyId\": {\"$numberLong\": \"6794045393576394753\"}}}}
	at com.mongodb.internal.connection.ProtocolHelper.getCommandFailureException(ProtocolHelper.java:175)
	at com.mongodb.internal.connection.InternalStreamConnection$2$1.onResult(InternalStreamConnection.java:390)
	at com.mongodb.internal.connection.InternalStreamConnection$2$1.onResult(InternalStreamConnection.java:376)
	at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback$MessageCallback.onResult(InternalStreamConnection.java:677)
	at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback$MessageCallback.onResult(InternalStreamConnection.java:644)
	at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:514)
	at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:511)
	at com.mongodb.connection.netty.NettyStream.readAsync(NettyStream.java:233)
	at com.mongodb.internal.connection.InternalStreamConnection.readAsync(InternalStreamConnection.java:511)
	at com.mongodb.internal.connection.InternalStreamConnection.access$1000(InternalStreamConnection.java:76)
	at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback.onResult(InternalStreamConnection.java:634)
	at com.mongodb.internal.connection.InternalStreamConnection$MessageHeaderCallback.onResult(InternalStreamConnection.java:619)
	at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:514)
	at com.mongodb.internal.connection.InternalStreamConnection$5.completed(InternalStreamConnection.java:511)
	at com.mongodb.connection.netty.NettyStream.readAsync(NettyStream.java:233)
	at com.mongodb.connection.netty.NettyStream.handleReadResponse(NettyStream.java:263)
	at com.mongodb.connection.netty.NettyStream.access$800(NettyStream.java:69)
	at com.mongodb.connection.netty.NettyStream$InboundBufferHandler.channelRead0(NettyStream.java:322)
	at com.mongodb.connection.netty.NettyStream$InboundBufferHandler.channelRead0(NettyStream.java:319)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1470)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1219)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1266)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:493)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:432)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:271)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:834)
```

and here is the corresponding code snippet

```
val limit = 100
  val readJournal: CurrentPersistenceIdsQuery = PersistenceQuery(system).readJournalFor[CurrentPersistenceIdsQuery](MongoReadJournal.Identifier)

  def getAll: Action[AnyContent] = Action.async {
    val futurePids: Future[immutable.Seq[String]] = readJournal.currentPersistenceIds()
      .take(limit)
      .runWith(Sink.seq)
      .recover({
        case e: Throwable =>
          logger.error("db empty?", e)
          immutable.Seq.empty[String]
      })
    val futureSum: Future[Int] = readJournal.currentPersistenceIds().map(_ => 1).runFold[Int](0)(_ + _)
    futurePids.zip(futureSum).map({
      case (pids, sum) => Ok(Json.obj("pids" -> pids, "total" -> sum))
    })
  }
```

I'm fairly new to scala and akka-persistence, but it looks like the `.take(limit)` happens just after the complete collection is queried. I didn't find a way to pass `allowDiskUse` as proposed in the error message in akka-persistence, only in the official mongo scala driver.

I'm using the following dependencies
```
val akkaVersion = "2.5.27"
libraryDependencies += "com.typesafe.akka" %% "akka-persistence" % akkaVersion
libraryDependencies += "com.typesafe.akka" %% "akka-persistence-query" % akkaVersion
libraryDependencies += "com.github.scullxbones" %% "akka-persistence-mongo-scala" % "2.3.2"
libraryDependencies += "org.mongodb.scala" %% "mongo-scala-driver" % "2.8.0"
```

Any ideas would be very appreciated!

**Comment by scullxbones** [comment]
Oof that's a lot of persistence ids:

> $group Operator and Memory

> The $group stage has a limit of 100 megabytes of RAM. By default, if the stage exceeds this limit, $group returns an error. To allow for the handling of large datasets, set the allowDiskUse option to true. This flag enables $group operations to write to temporary files. For more information, see the db.collection.aggregate() method and the aggregate command.

I double checked the code and I *think* the `$group` should be the smallest possible size.

I think the options to fix the problem are:

- Somehow expose `allowDiskUse = true` override in the configuration
- Figure out another approach to find distinct persistence ids

I have a feeling `allowDiskUse` is extremely slow.  Have you tried creating a local build of the driver and hardcoding this value?   Would be good for a feasibility check.

Also I wonder, what changed upon switching database providers?  Was it a simple mongodump and restore?  I haven't used Atlas yet.  Or could it be that Atlas runs on a downsized max grouping memory?  I don't see any way to override 100 MB, but they could create a special build just for Atlas...

**Comment by scullxbones** [comment]
I found an idea for a lower memory use approach with trade-offs that would get rid of this error:

[Stack overflow](https://stackoverflow.com/a/55491963) 
- Basically replace the `$group` with `$sort`, and use primitives in `akka-stream` to discard successive duplicates.  i.e. Stateful graph stage that does a `.takeWhile( <duplicate> ) > /dev/null`
- I *think* the tradeoff will be in network speed/bandwidth between app and database and/or CPU on the application side
  - I don't know how fat the network pipe is that Atlas provides, so this may also be slow

Another idea is to keep a materialized copy of distinct `persistenceIds`, with associated data sync and migration problems, including the starting problem.

There's also `allowDiskUse` of course.

**Comment by clteich** [comment]
I tried creating a local build and it seemed to work locally, or at least now I run into the configured socket timeout because the slow query. I just hard-coded `allowDiskUse` in ScalaDriverPersistenceReadJournaller.scala as follows:

```
object CurrentPersistenceIds {
  def source(driver: ScalaMongoDriver): Source[String, NotUsed] = {
    Source.future(driver.journalCollectionsAsFuture)
      .mapConcat(identity)
      .flatMapConcat(_.aggregate(project(include(PROCESSOR_ID)) :: group(s"$$$PROCESSOR_ID") :: Nil).allowDiskUse(true).asAkka)
      .map { doc => Option(doc.getString("_id").getValue).getOrElse("") }
      .filterNot(_.isEmpty)
  }
}
```

Unfortunately I didn't find a way to override the configuration for all aggregations.

> Also I wonder, what changed upon switching database providers? Was it a simple mongodump and restore? I haven't used Atlas yet. Or could it be that Atlas runs on a downsized max grouping memory? I don't see any way to override 100 MB, but they could create a special build just for Atlas...
Yeah, it was just a dump and restore. I'm also in contact with our support there and asked about if there's a correlation between cluster tier and max grouping memory. I'll post the answer here. Otherwise both databases have almost the same RAM and vCPUs.

**Comment by clteich** [comment]
We'll use a different migration strategy to workaround our issues.

**Comment by scullxbones** [comment]
Was `allowDiskUse` too slow?  What ended up happening?

**Comment by clteich** [comment]
We ran into a separate problem with the de/serialization of actor states after migrating the data. We decided against data migration and we'll run both environment in parallel for a while, so the old one "dries out", i.e. until the actor system handled all off the remaining, possible messages.

I think it would still be a worthwhile improvement to make `allowDiskUse` configurable. Maybe just read the config-value in the code snippet posted above.
